File: tests/memtest/mem_test.cpp
// Memory test for the azul C++ binding. See tests/memtest/README.md.
//
// The harness measures RSS/gdb externally; this file just exercises the
// create/consume/DROP paths in a loop and exits 0. No event loop (App::run
// needs a display and hangs headless). The RAII wrappers (~AppConfig, ~Dom,
// ~App) call the Az*_delete functions on scope exit.
//
// Build (matches examples/cpp/cpp11/hello-world.cpp):
//   g++ -std=c++11 -o mem_test mem_test.cpp -lazul -L<target/release> \
//       -I<repo>/target/codegen

#include "azul11.hpp"
#include <cstdio>
#include <cstdlib>

using namespace azul;

struct MyDataModel { uint32_t counter; };

int main() {
    const char* n_env = std::getenv("AZ_MEMTEST_N");
    long n = n_env ? std::strtol(n_env, nullptr, 10) : 200000;
    if (n < 0) n = 0;

    // 1. The consume-by-value DROP path: App::create consumes RefAny +
    //    AppConfig; ~App destroys it. No run() -> no window.
    {
        MyDataModel model = { 5 };
        RefAny data = RefAny::create(std::move(model));
        App app = App::create(std::move(data), AppConfig::create());
    } // ~App here

    // 2. Leak loop: create/destroy droppable objects N times (RAII deletes).
    for (long i = 0; i < n; i++) {
        AppConfig cfg = AppConfig::create();
        Dom dom = Dom::create_body();
        (void)cfg;
        (void)dom;
    } // ~AppConfig + ~Dom each iteration

    std::printf("memtest cpp OK (N=%ld)\n", n);
    return 0;
}

File: tests/memtest/mem_test.lua
-- Memory test for the azul Lua (LuaJIT ffi) binding. See tests/memtest/README.md.
--
-- The harness measures RSS/gdb externally; this file just exercises the
-- create/consume/DROP paths in a loop and exits 0. No event loop (app:run
-- needs a display and hangs headless). Native objects carry an ffi.gc
-- finalizer, so the loop drops references and nudges collectgarbage.
--
-- Run (matches examples/lua/hello-world.lua):
--   LD_LIBRARY_PATH=. luajit mem_test.lua

local azul = require('azul')

local N = tonumber(os.getenv('AZ_MEMTEST_N')) or 200000

-- 1. The consume-by-value DROP path: App.create consumes the refany +
--    AppConfig. Build it without running (no window), then let it drop.
do
    local data = azul.refany_create({ counter = 5 })
    local app = azul.App.create(data, azul.AppConfig.create())
    app = nil
end
collectgarbage('collect')

-- 2. Leak loop: create/destroy droppable objects N times. ffi.gc frees the
--    native memory on collection, so drop the refs and nudge the collector.
for i = 1, N do
    local cfg = azul.AppConfig.create()
    local dom = azul.Dom.create_body()
    cfg = nil
    dom = nil
    if i % 16384 == 0 then collectgarbage('collect') end
end
collectgarbage('collect')

print(string.format('memtest lua OK (N=%d)', N))
os.exit(0)
